crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+            }
+        }
+
+        // We assume here that the first constructor is the "correct" type. Since we
+        // only care about the "type" of the constructor (i.e. if it is a bool we
+        // don't care about the value), this assumption should be valid as long as
+        // the match statement is well formed. We currently uphold this invariant by
+        // filtering match arms before calling `is_useful`, only passing in match arms
+        // whose type matches the type of the match expression.
+        match &used_constructors.first() {
+            Some(constructor) if all_constructors_covered(&cx, constructor, &used_constructors) => {
+                // If all constructors are covered, then we need to consider whether
+                // any values are covered by this wildcard.
+                //
+                // For example, with matrix '[[Some(true)], [None]]', all
+                // constructors are covered (`Some`/`None`), so we need
+                // to perform specialization to see that our wildcard will cover
+                // the `Some(false)` case.
+                let mut constructor = None;
+                for pat in matrix.heads() {
+                    if let Some(c) = pat_constructor(cx, pat)? {
+                        constructor = Some(c);
+                        break;
+                    }
+                }
+
+                if let Some(constructor) = constructor {
+                    if let Constructor::Enum(e) = constructor {
+                        // For enums we handle each variant as a distinct constructor, so
+                        // here we create a constructor for each variant and then check
+                        // usefulness after specializing for that constructor.
+                        let mut found_unimplemented = false;
+                        for constructor in
+                            cx.db.enum_data(e.parent).variants.iter().map(|(local_id, _)| {
+                                Constructor::Enum(EnumVariantId { parent: e.parent, local_id })
+                            })
+                        {
+                            let matrix = matrix.specialize_constructor(&cx, &constructor)?;
+                            let v = v.expand_wildcard(&cx, &constructor)?;
+
+                            match is_useful(&cx, &matrix, &v) {
+                                Ok(Usefulness::Useful) => return Ok(Usefulness::Useful),
+                                Ok(Usefulness::NotUseful) => continue,
+                                _ => found_unimplemented = true,
+                            };
+                        }
+
+                        if found_unimplemented {
+                            Err(MatchCheckNotImplemented)
+                        } else {
+                            Ok(Usefulness::NotUseful)
+                        }
+                    } else {
+                        let matrix = matrix.specialize_constructor(&cx, &constructor)?;
```

Comment:
it seems wrong to not specialize by all constructors (i.e. `true` and `false` for booleans) here. how about `(false, _), (true, false), (_, true)` as a test case?

(I think it'd make sense to have a function that returns all constructors for a type, like `all_constructors` in rustc, and use that instead of handling enums specially here.)

Reply:
Added `all_constructors` as an impl on `Constructor`, fixed the behavior here, removed the special casing for enums, and added the suggested test.